esp/services/ws_dfu/ws_dfuService.cpp

```diff
@@ -6118,7 +6104,289 @@ bool CWsDfuEx::onDFUFileAccess(IEspContext &context, IEspDFUFileAccessRequest &r
+    }
+    return true;
+}
+
+const char *CWsDfuEx::createWsDFUCreateGroupName(StringArray &locations, StringBuffer &groupName)
+{
+    if (locations.length() < 1)
+        return nullptr;
+
+    StringBuffer buf;
+    ForEachItemIn(i, locations)
+        buf.append(locations.item(i)).append(";");
+    groupName.set("wsdfucreate").append(hashc((unsigned char*) buf.str(), buf.length(), 0));
+    return groupName.str();
+}
+
+IGroup *CWsDfuEx::queryDFUFileIGroup(StringBuffer &groupName, ClusterType clusterType, StringArray &locations)
+{
+    GroupType groupType;
+    StringBuffer basedir;
+    Owned<IGroup> group = queryNamedGroupStore().lookup(groupName.str(), basedir, groupType);
+    if (group)
+    {
+        if (groupType == grp_roxie)
+            throw MakeStringException(ECLWATCH_INVALID_INPUT, "IGroup for roxie not allowed: %s.", groupName.str());
+
+        ESPLOG(LogMax, "Found DFUFileIGroup %s", groupName.str());
+        return group.getClear();
+    }
+
+    if (locations.length() < 1)
+        throw MakeStringException(ECLWATCH_INVALID_INPUT, "PartLocations not defined.");
+
+    StringBuffer defaultDir;
+    if (!getConfigurationDirectory(nullptr, "data", (clusterType == HThorCluster) ? "hthor" : "thor", groupName, defaultDir))
+        throw MakeStringException(ECLWATCH_INVALID_INPUT, "Failed to get ConfigurationDirectory: %s.", groupName.str());
+
+    SocketEndpointArray epa;
+    ForEachItemIn(i, locations)
+    {
+        SocketEndpoint ep(locations.item(i));
+        if (ep.isNull())
+            throw MakeStringException(ECLWATCH_INVALID_INPUT, "Invalid location '%s'.", locations.item(i));
+        epa.append(ep);
+    }
+    group.setown(createIGroup(epa));
+    queryNamedGroupStore().add(groupName.str(), group, false, defaultDir.str(), (clusterType == HThorCluster) ? grp_hthor : grp_thor);
+    ESPLOG(LogMin, "DFUFileIGroup %s added", groupName.str());
+    return group.getClear();
+}
+
+void CWsDfuEx::exportRecordDefinitionBinaryType(const char *recordDefinition, MemoryBuffer &layoutBin)
+{
+    MultiErrorReceiver errs;
+    Owned<IHqlExpression> expr = parseQuery(recordDefinition, &errs);
+    if (errs.errCount() > 0)
+    {
+        StringBuffer errorMsg;
+        throw MakeStringException(ECLWATCH_INVALID_INPUT, "Failed in parsing ECL %s: %s.", recordDefinition, errs.toString(errorMsg).str());
+    }
+    if (!expr)
+        throw MakeStringException(ECLWATCH_INVALID_INPUT, "Failed in parsing ECL: %s.", recordDefinition);
+    
+    if (!exportBinaryType(layoutBin, expr, false))
+        throw MakeStringException(ECLWATCH_INVALID_INPUT, "Failed in exportBinaryType.");
+}
+
+IFileDescriptor *CWsDfuEx::createDFUDescriptor(IEspContext &context, const char *fileName, ClusterType clusterType,
+    const char *jobId, const char *userId, StringArray &locations, StringBuffer &groupName)
+{
+    Owned<IGroup> group = queryDFUFileIGroup(groupName, clusterType, locations);
```

Comment:
As per other comment createDFUDescriptor doesn't look like it should belong in esp/wsdfu, but as a utility method of dafdesc.cpp or similar general use..

In which case the IGroup should be passed in, and therefore the line above called outside of this function.